### PR TITLE
Local Processor Interface

### PR DESCRIPTION
**Pull request by nvidianz**
### Description
1. LocalProcessor interface for C++ plugins.
2. Implemented a new mock plugin to test the LocalProcessor interface.
### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: integration/xgboost/processor/src/dam/dam.cc [hunk 1]
```diff
@@ -17,62 +17,136 @@
 #include <cstring>
 #include "dam.h"
 
+
+void print_hex(const uint8_t *buffer, int size) {
+    if (size )
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/dam/dam.cc:22]: Format =>: if (size)

File: integration/xgboost/processor/src/include/local_processor.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    Buffer encrypted_gh_;
+    std::vector<double> *histo_{nullptr};
+    std::vector<uint32_t> cuts_;
+    std::vector<int> slots_;
+    bool print_timing_ = false;
+    bool debug_ = false;
+    bool dam_debug_ = false;
+
+ public:
+    void Initialize(bool active, std::map<std::string, std::string> params) override;
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/include/local_processor.h:38]: Please add comments to explain methods of this API.

File: integration/xgboost/processor/src/include/local_processor.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    void Shutdown() override;
+
+    void FreeBuffer(void *buffer) override;
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/include/local_processor.h:42]: Explain what this is and when it is to be called.

File: integration/xgboost/processor/src/nvflare-plugin/plugin_main.cc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cstring>
+#include "nvflare_processor.h"
+#include "local_mock.h"
+
+extern "C" {
+
+processing::Processor *LoadProcessor(char *plugin_name) {
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/plugin_main.cc:22]: What is this plugin_main used for?

File: integration/xgboost/processor/src/nvflare-plugin/util.cc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * limitations under the License.
+ */
+#include "util.h"
+#include <iostream>
+#include <set>
+#include <algorithm>
+
+const double kScaleFactor = 1000000.0;
+
+std::vector<std::pair<int, int>> distribute_work(size_t num_jobs, size_t num_workers) {
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/util.cc:23]: What is it for?

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Looks good in general. See my comments for improvement.  Main thing is lack of explanation/comments about how the API is to be used/implemented.

**Review by YuanTingHsieh — COMMENTED** [review]
Mostly good, one question

**Comment by nvidianz** [comment]
The change is merged with other PRs.

**Pull request closed without merging**